### PR TITLE
Make app loading offline first

### PR DESCRIPTION
**Pull request by garethbowen**
# Description

Modify purging to run in two stages: cache IDs in the background, and purge locally on bootstrap.
Remove blocking on session check from angular startup.
Add test for startup on turtle connection.
Refactoring to make bootstrap code more readable.

medic/cht-core#7492

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

File: tests/e2e/login/purge.spec.js [hunk 1]
```diff
@@ -1,315 +0,0 @@
...
-      const db = window.CHTCore.DB.get();
-      db.get('_local/purgelog')
-        .then(doc => callback(doc))
-        .catch(err => callback(err));
-    }));
-  };
-
-  const restartSentinel = () => utils.stopSentinel().then(() => utils.startSentinel());
-
-  it('Logging in as a restricted user with configured purge rules should not download purged docs', async () => {
```
Comment (dianabarsan) [inline tests/e2e/login/purge.spec.js:242]: Don't you think there's still value in having this around? it actually tests which docs get purged, with different purge rules, while the other test does not.

**Review by dianabarsan — APPROVED** [review]
Awesome work! 
I just have a couple of questions inline, but no blockers.

File: webapp/src/js/bootstrapper/purger.js [hunk 2]
```diff
@@ -1,149 +1,120 @@
...
+    .then(() => {
+      return getPurgeLog(localDb)
+        .then(log => {
+          if (!log.to_purge || !log.to_purge.length) {
+            return;
+          }
+          emit('start');
+          return batchedPurge(log.to_purge)
+            .then(() => writePurgeLog(localDb, totalPurged, userCtx))
+            .then(() => clearToPurgeList(localDb))
```
[the remark this answers is not in the document]
Comment (garethbowen) [reply webapp/src/js/bootstrapper/purger.js:76]: I've added this. I still update the purgelog history after all batches have run, but now I'm not sure. Should each batch cause a new history item?

File: tests/e2e/login/purge.spec.js (hunk 1 above)
Comment (garethbowen) [reply tests/e2e/login/purge.spec.js:242]: I'm not sure I follow. The purge.wdio-spec.js e2e test has one configured purge function, runs purge on the server, refreshes the browser, makes sure the right docs are purged, then changes the purge function, runs purge on the server, refreshes the browser, and makes sure additional docs are purged. What is it missing?

Comment (dianabarsan) [reply tests/e2e/login/purge.spec.js:242]: The one thing it's missing is all the assertions on purge history contents. 
Maybe we could add assertions for `to_purge` after sync and add assertions for the purge history entry in this test?

File: webapp/src/js/bootstrapper/purger.js [hunk 3]
```diff
@@ -32,10 +32,12 @@ const appendToPurgeList = (localDb, ids) => {
     });
 };
 
-const clearToPurgeList = (localDb) => {
+const removeFromToPurgeList = (localDb, docIds) => {
   return getPurgeLog(localDb)
     .then(log => {
-      log.to_purge = [];
+      log.to_purge = log.to_purge.filter(id => {
+        return docIds.indexOf(id) === -1;
```
Comment (dianabarsan) [inline webapp/src/js/bootstrapper/purger.js:39]:
```suggestion
        return !docIds.includes(id);
```
No magic numbers :D

**Review by dianabarsan — COMMENTED** [review]
On history, if the process indeed gets interrupted, and we only write history at the end, we would indeed lose the delete count of docs that were purged before interruption.

**Comment by garethbowen** [comment]
> On history, if the process indeed gets interrupted, and we only write history at the end, we would indeed lose the delete count of docs that were purged before interruption.

Yeah I agree. So I should change it to write history after each page? It seems like a bad idea because the history only stores 10 items, so with 1k docs in the to_purge list it'll completely rewrite the history. It's also possible to have >1k depending on how previous runs went which means there will be no tracking for some changes. I could increase the history length, or increase the page size (why not 1k?)?

File: webapp/src/js/bootstrapper/purger.js (hunk 3 above)
Comment (garethbowen) [reply webapp/src/js/bootstrapper/purger.js:39]: I'm so gun-shy about newfangled features in native code for 3.x. Are you sure that'll work in all browsers?

**Comment by dianabarsan** [comment]
> or increase the page size (why not 1k?)?

That would simplify things! The only thing I'd be worried about is some stupid indexeddb crash...

File: webapp/src/js/bootstrapper/purger.js (hunk 3 above)
Comment (dianabarsan) [reply webapp/src/js/bootstrapper/purger.js:39]: yea, it's added in Chrome 47 https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/includes#browser_compatibility

Comment (garethbowen) [reply webapp/src/js/bootstrapper/purger.js:39]: Done.

File: tests/e2e/login/purge.spec.js (hunk 1 above)
Comment (garethbowen) [reply tests/e2e/login/purge.spec.js:242]: I've added the extra assertions.

**Comment by garethbowen** [comment]
I increased the history items to 20 and now update the history each batch not at the end.

**Pull request merged**